doc/devguide/extending/detect/app-layer-frames.rst

```diff
@@ -0,0 +1,190 @@
+
+As the example rule shows, this allows for easier to write, more human-readable rules, with the benefit that they are also less costly for the engine to inspect.
+
+Rule Writing
+============
+
+An advantage of this feature is that there is a single rule keyword for the frame record: ``record``.
+This keyword takes an argument to specify the per protocol record type::
+
+    record:<app proto name>.<specific record name>
+
+Examples::
+
+    tls.pdu
+    tls.hdr
+    smb.smb2.smb2.hdr
+    smb.smb3.data
+
+Rules will look something like this::
+
+    alert tls any any -> any any (flow:to_server; record:tls.pdu; \
+         content:"|16 03|"; startswith; sid:1;)
+    alert tls any any -> any any (flow:to_server; record:tls.hdr; \
+         content:"|16 03|"; startswith; sid:2;)
+    alert tls any any -> any any (flow:to_server; record:tls.data; \
+         content:"|aa aa aa aa|"; startswith; sid:3;)
+    alert smb any any -> any any (flow:to_server; record:smb.nbss.pdu; \
+         content:"|fe|SMB"; offset:4; depth:4; sid:25;)
+    alert smb any any -> any any (flow:to_client; record:smb.nbss.pdu; \
+         content:"|fe|SMB"; offset:4; depth:4; sid:26;)
+    alert smb any any -> any any (flow:to_server; record:smb.smb1.pdu; content:"|ff|SMB"; sid:21;)
+
+
+Implementation & API Callbacks
+==============================
+
+The application layer parser exposes records it supports to the detect engine, by tagging them as they're parsed. The rest works automatically.
+
+A code example from the `SMB` decoder (`rust/src/smb/smb.rs`) - code excerpt from `PR #6684 <https://github.com/OISF/suricata/pull/6684/commits/2921c6c81f2ab72adbab146670810dc6d869db52>`_:
+
+    .. code-block:: rust
+
+         pub fn parse_tcp_data_ts<'b>(&mut self, flow: *const Flow, stream_slice: &StreamSlice) 
+         -> AppLayerResult
+         {
+            let mut cur_i = stream_slice.as_slice(); // pointer to current input
+               .
+               while cur_i.len() > 0 {
+                 match parse_nbss_record(cur_i) {
+                 .
+                 .
+                   let (_nbss_pdu, _nbss_pdu_id) = applayer_new_frame_ts(flow, stream_slice, cur_i,
+                           nbss_hdr.length as i32 + 4, SMBFrameType::NBSS as u8);
+                   let (_nbss_hdr_frame, _nbss_frame_id) = applayer_new_frame_ts(flow, stream_slice, 
+                           cur_i, 4, SMBFrameType::NBSSHdr as u8);
+                   let (_nbss_data_frame, _nbss_data_frame_id) = applayer_new_frame_ts(flow, 
+                           stream_slice, &cur_i[4..], nbss_hdr.length as i32, SMBFrameType::NBSSData as u8);
+                 .
+                 .
+                 }
+            }
+         }
```

Comment:
See if I manage to use the `literalinclude`directive to show real code snippet here.

As in: https://github.com/OISF/suricata/pull/5666/commits/28bc166315822eca58fe7eb96302a9245a5572d4
and https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude